File: 63. Unique Paths II/solution63.py
from typing import List

class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        # print(numpy.matrix(obstacleGrid))
        ROWS, COLS = len(obstacleGrid), len(obstacleGrid[0])
        obstacleGrid = [
            [float("+inf") if square == 1 else 1 for square in row]
            for row in obstacleGrid
        ]
        is_after_obstacle = False
        for col in range(COLS):
            if obstacleGrid[0][col] == float("+inf"):
                is_after_obstacle = True
            elif is_after_obstacle:
                obstacleGrid[0][col] = float("+inf")
        is_after_obstacle = False
        for row in range(ROWS):
            if obstacleGrid[row][0] == float("+inf"):
                is_after_obstacle = True
            elif is_after_obstacle:
                obstacleGrid[row][0] = float("+inf")
        # print(numpy.matrix(obstacleGrid))
        for row in range(1, ROWS):
            for col in range(1, COLS):
                if obstacleGrid[row][col] == float("+inf"):
                    continue
                top_sq = obstacleGrid[row - 1][col]
                left_sq = obstacleGrid[row][col - 1]
                if top_sq == float("+inf") and left_sq == float("+inf"):
                    obstacleGrid[row][col] = float("+inf")
                else:
                    obstacleGrid[row][col] = (
                        top_sq if top_sq != float("+inf") else 0
                    ) + (left_sq if left_sq != float("+inf") else 0)
                    pass
        # print(numpy.matrix(obstacleGrid))
        ans = obstacleGrid[-1][-1] 
        return ans if ans != float("+inf") else 0

ans = Solution().uniquePathsWithObstacles([[0, 0, 0], [0, 1, 0], [0, 0, 0]])
assert ans == 2
ans = Solution().uniquePathsWithObstacles([[0, 0, 0], [0, 1, 1], [0, 0, 0]])
assert ans == 1
ans = Solution().uniquePathsWithObstacles([[0,1],[0,0]])
assert ans == 1
ans = Solution().uniquePathsWithObstacles([[0,1],[1,0]])
assert ans == 0
ans = Solution().uniquePathsWithObstacles([[0,1],[1,0]])
assert ans == 0

File: 63. Unique Paths II/solution63.c
#include <stdbool.h>
#include <stdlib.h>
#include <assert.h>

#include <stdbool.h>
#include <stdlib.h>
#include <assert.h>
#include <stdio.h>

int uniquePathsWithObstacles(int **grid, int ROWS, int *COLS)
{
    grid[0][0] = (grid[0][0] << 1) + !grid[0][0];
    for (size_t col = 1; col < *COLS; ++col)
    {
        if (grid[0][col] == 1)
            grid[0][0] |= 4;
        grid[0][col] = grid[0][col] == 0 && !(grid[0][0] & 6) ? 1 : 0;
    }
    for (size_t row = 1; row < ROWS; ++row)
    {
        if (grid[row][0] == 1)
            grid[0][0] |= 8;
        grid[row][0] = grid[row][0] == 0 && !(grid[0][0] & 10) ? 1 : 0;
    }
    grid[0][0] &= 1;


    for (size_t row = 1; row < ROWS; ++row)
    {
        for (size_t col = 1; col < *COLS; ++col)
        {
            grid[row][col] = grid[row][col] == 0 ? grid[row - 1][col] + grid[row][col - 1] : 0;
        }
    }
    return grid[ROWS - 1][*COLS - 1];
}

int main()
{
    int ROWS = 2, COLS = 1;
    int **grid = (int **)malloc(sizeof(int *) * ROWS);
    for (size_t row = 0; row < ROWS; ++row)
    {
        grid[row] = malloc(sizeof(int) * COLS);
    }
    grid[0][0] = 1;
    grid[1][0] = 0;
    int ans = uniquePathsWithObstacles(grid, ROWS, &COLS);
    printf("ans == %d\n", ans);
    assert(ans == 0);
    for (size_t row = 0; row < ROWS; ++row)
    {
        free(grid[row]);
    }
    free(grid);

    return 0;
}